feat(view): Add viewer rewrite to monorepo

mcous opened the pull request:
## overview

This PR is part 3 of mcous/tracespace#1, and adds the new viewer app to the monorepo. It's written in Typescript and still needs tests, but this (I think) is feature-complete enough to launch.

Preview at: <https://staging.tracespace.io/view>

Still TODO in follow up PR(s):

- Contributing documentation updates
- Set up CI deploy
- Set up Greenkeeper
- Set up offline support
- Component consolidation and simplification
- Way more tests

## changelog

Improvements from the code living at [tracespace/viewer](https://github.com/tracespace/viewer):

- Uses the latest tracespace libraries, which fixes tracespace/viewer#15
- Written in React instead of a random dead framework
- Actually works with Firefox when:
    - scrolling side panels (closes tracespace/viewer#3)
    - scrolling to zoom with a mouse instead of a trackpad (closes tracespace/viewer#10)
- Can drop files anywhere in the app to open them (closes tracespace/viewer#5)
- Files that don't render have an indicator and tooltip (closes tracespace/viewer#6)
- There's a download button to save all the SVG renders (closes tracespace/viewer#9)
- Link to GitHub repo from site (closes tracespace/viewer#11)
- Supports opening zip files (closes tracespace/viewer#12)
- Can easily isolate layers in layer view with shift+click on visibility toggle (closes tracespace/viewer#13)
- More obvious first use instructions + a default render to try (Arduino) ready to go (closes tracespace/viewer#18)
- All relevant render options can be set by the user (closes tracespace/viewer#19)

## try it out

```shell
# install dependencies
cd tracespace
npm i

# launch dev server on localhost:8080
# also launches pcb-stackup, pcb-stackup-core, gerber-to-svg dev servers
npm start

# build production assets and serve them on localhost:9090
npm run build
npm run serve
```

kasbah commented:
So cool! It's looking great, as always really impressed by the code quality and organization too (makes my projects look like a pile of random source files). 

Noticed some issues :

1. I expect to be able to click and drag the bottom bar to be able to zoom but doesn't seem to happen
2. On Chrome (Version 72.0.3626.109 64-bit Linux) zooming doesn't perform well, I thought it didn't work at all at first
4. On Chrome when trying to download SVGs of the [ulx3s](https://kitspace.org/boards/github.com/emard/ulx3s/) it doesn't work due to this  `Error: <circle> attribute r: Expected length, "NaN"` (it seems to have some issue with a "drawing" layer which appears as huge circles as well). 
3. Maybe you want to style this scroll bar on the left -- it seems to be the default system style and could easily clash? 
![image](https://user-images.githubusercontent.com/206854/53178490-da70fa00-35e9-11e9-9fbd-71fb76d101a2.png)

5. URL upload input should delete the default, i.e. be empty, when you focus on it

6. URL upload doesn't seem to work for Kitspace, e.g. https://kitspace.org/boards/github.com/emard/ulx3s/ulx3s-6466905-gerbers.zip due to CORS :(, I probably need to enable it on the Kitspace server but it will probably be a common issue, maybe a little server side proxy/downloader would be best.

mcous commented on typings/raf-throttle.d.ts:
```diff
@@ -0,0 +1,15 @@
+// Type definitions for raf-throttle 1.0
```
I think I removed this dependency, so should probably remove this typings file, too

mcous commented on typings/nanologger.d.ts:
```diff
@@ -0,0 +1,44 @@
+// Type definitions for nanologger 1.3
```
Removed this dependency

mcous commented on .eslintrc.js:
```diff
@@ -14,7 +14,17 @@ module.exports = {
+  ],
+  plugins: ['react-hooks'],
+  rules: {
+    'react-hooks/rules-of-hooks': 'error',
```
This rule is triggering in non-React code. Consider scoping it

mcous commented:
Alright @kasbah how would you feel if I merged this now? Regarding your feedback:

1. Implemented: 8c6230f

2. Chrome on Linux (on my machine) was occasionally creating wheel events with `deltaY: -0`. This threw off an overly naive `Math.sign` comparison in the zoom throttler. Fixed with 6807f41

3. Wasn't able to reproduce this on any of my computers / VMs with Chrome and Firefox. Can hopefully address if we get a repro

4. I added some really basic styling using the [CSS scrollbars](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Scrollbars) standard proposal properties. Only Firefox supports them, but also Chrome on Lniux doesn't seem to mess with the default scrollbars like Firefox does, so I think that's all I really feel like doing for now. 622bdd1

5. I still like the ability to quickly load an example (see tracespace/viewer#18), but I did add a select-all-text-on-click to the URL input field to make it easier to wipe the field out with your own url

6. Still don't feel like setting up any backend infrastructure to support servers not set up for CORS just yet, but I did implement a very simple error toast so at least you know _something_ went wrong. Future improvements here would be actual user-actionable error messages based on known failure modes. 22c42e0

    ![2019-02-24 11 33 27](https://user-images.githubusercontent.com/2963448/53302038-0ed1f980-3828-11e9-9a9d-e4bd5a0437e4.gif)

I just pushed everything above to staging. Todos listed in PR description (or maybe just the CI one) are still gating for the actual release I think

kasbah commented:
I feel good about merging in general since it's all brand new! Just starting to read the source to pick up some tricks.  

>2. Chrome on Linux (on my machine) was occasionally creating wheel events with `deltaY: -0`. This threw off an overly naive `Math.sign` comparison in the zoom throttler. Fixed with [6807f41](https://github.com/tracespace/tracespace/commit/6807f41a3eaf5019ad1003d914171366f8f69298)
>3. Wasn't able to reproduce this on any of my computers / VMs with Chrome and Firefox. Can hopefully address if we get a repro

So I am still experiencing the zoom issues on Chrome with my Thinkpad x230 Touchpad on Ubuntu 16.04. On Firefox it's still fine, though it does feel a bit like it's "jumpy" on both actually since there only seem to be a few zoom positions to select from. Is there a reason for that?

mcous commented:
> So I am still experiencing the zoom issue on Chrome with my Thinkpad x230 Touchpad on Ubuntu 16.04

Shoot. I've just pushed another change to the zoom throttling to this branch and staging. I'll also try to spin up a Ubuntu 16.04 VM (My machines are 18.04 currently) to test

> it does feel a bit like it's "jumpy" on both actually since there only seem to be a few zoom positions to select from. Is there a reason for that?

Yeah still trying to optimize this. For a while with this I was trying to do a continuous zoom, but even on my relatively powerful machine I was finding the "continuous" experience to be really janky. I think this speaks to a need to really concentrate on simplifying the SVG renders themselves so the browser can do less work (e.g. #80), but for now...

The zoom is descretized into 17 ticks. This _really_ simplifies some stuff (because zoom levels, both linear and logarithmic, can be pre-calculated), but I think that the zoom bar needs to do a better job of indicating that the zoom levels are discrete positions.

Longer term, I think a combination of increasing the number of ticks and also getting CSS transitions involved for the zoom itself will improve things, and possibly even trick the mind into thinking the zoom is continuous. At the moment though, the aforementioned render performance issues mean zoom transitions are _also_ janky, so I left them out.

I figure discrete zoom levels is the least bad option at the moment 🙃

mcous commented on apps/view/src/FileList/FileItem.tsx:
```diff
@@ -0,0 +1,35 @@
+  const {id, side, type, converter, filename, color, showFilenames} = props
+
+  return (
+    <li className="flex items-center h2 mb2 pl3 br3 overflow-hidden bg-white shadow">
```
Move styles to constants for legibility

mcous commented on apps/view/src/FileList/ShowFilenamesButton.tsx:
```diff
@@ -0,0 +1,24 @@
+
+  return (
+    <Button
+      className="absolute left-1 bottom-2 flex items-center f5"
```
Move styles to constants

mcous commented on apps/view/src/FileList/SideItem.tsx:
```diff
@@ -0,0 +1,26 @@
+
+  return (
+    <li>
+      <h3 className="mv2 pl3 lh-title f5 b">{label}</h3>
```
"

mcous commented on apps/view/src/FileList/SideList.tsx:
```diff
@@ -0,0 +1,52 @@
+  return (
+    <div
+      className={cx(
+        'absolute left-0 top-5 bottom-5 overflow-y-hidden w-animate',
```
"

mcous commented on apps/view/src/LoadFiles/index.tsx:
```diff
@@ -0,0 +1,85 @@
+  handleUrl: (url: string) => void
+}
+
+export default function LoadFiles(props: LoadFilesProps): JSX.Element {
```
This component is too complex. Break it up

mcous commented on apps/view/src/Nav/Footer.tsx:
```diff
@@ -0,0 +1,37 @@
+
+export default function Footer(): JSX.Element {
+  return (
+    <footer className="fixed right-1 bottom-1 w-third tr">
```
Styles to constants for consistency

mcous commented on apps/view/src/Nav/OpenFileDrawer.tsx:
```diff
@@ -0,0 +1,67 @@
+  return (
+    <Drawer title={TITLE} open={open} close={close}>
+      <Label>
+        <span className="mr-auto">{UPLOAD_FROM_COMPUTER_LABEL}</span>
```
Missed this style, move to const

mcous commented on apps/view/src/Nav/OpenFileDrawer.tsx:
```diff
@@ -0,0 +1,67 @@
+    <Drawer title={TITLE} open={open} close={close}>
+      <Label>
+        <span className="mr-auto">{UPLOAD_FROM_COMPUTER_LABEL}</span>
+        <input className="clip" type="file" onChange={handleFiles} multiple />
```
Probably should write a `HiddenInput` component

mcous commented on apps/view/src/Nav/OpenFileDrawer.tsx:
```diff
@@ -0,0 +1,67 @@
+
+const UPLOAD_FROM_COMPUTER_LABEL = 'upload files from your computer'
+const ENTER_URL_LABEL = 'enter the URL of a ZIP archive'
+const ENTER_URL_PLACEHOLDER = 'https://tracespace.io/view/arduino-uno.zip'
```
Need to replace this constant with the same logic that's used in the welcome screen URL input

mcous commented on apps/view/src/Nav/index.tsx:
```diff
@@ -0,0 +1,43 @@
+  const show = !loading && board !== null
+
+  return (
+    <nav className="flex items-start justify-between relative w-100 h3">
```
Styles up

mcous commented on apps/view/src/RenderWorker/models.ts:
```diff
@@ -0,0 +1,302 @@
+const SCALE_MM_TO_IN = 1 / 25.4
+const SCALE_IN_TO_MM = 25.4
+
+const getPcbStackup = import('pcb-stackup').then(m => m.default)
```
I think I've unintentionally missed a `() =>` here. This still works, but the chunk is no longer loaded lazily; the import is started as soon as this file is run. I don't think that behavior is what I intended

mcous commented on apps/view/src/db/index.ts:
```diff
@@ -0,0 +1,129 @@
+        .then(() => db.layers.orderBy('sourceId').uniqueKeys())
+        .then(sourceIdsToKeep => {
+          const sourceIdsToDelete = sourceIds.filter(
+            id => sourceIdsToKeep.indexOf(id) === -1
```
Prefer `Array.includes` I think. Already loading the polyfill so 🤷‍♂️

mcous commented on apps/view/src/images/types.d.ts:
```diff
@@ -0,0 +1,9 @@
+declare module '*.png' {
+  const href: string
+  export = href
```
I think I should prefer `export default` for ES6 modules (which is what webpack will be treating these as)

mcous commented on apps/view/src/layers.ts:
```diff
@@ -0,0 +1,146 @@
+  SIDE_BOTTOM,
+  SIDE_INNER,
+  SIDE_ALL,
+} from 'whats-that-gerber'
```
TODO: if we can somehow only import the contants (wtg is CommonJS, so no tree-shaking) we could get a small but meaningful size reduction on this chunk

mcous commented on apps/view/src/logger.ts:
```diff
@@ -0,0 +1,52 @@
+    ? (msg, ...meta) => console[lvl](`${lvl}: ${msg}`, ...meta)
+    : () => {}
+
+export function useLogger(): Logger {
```
I'm not sure this hook is behaving as desired; I can't enable debug logs in production

mcous commented on apps/www/README.md:
```diff
@@ -0,0 +1,5 @@
+
+> https://tracespace.io
+
+This page has not yet been build. Redirects to [/view](../view) for now.
```
`s/build/built/`

mcous commented on typings/viewbox.d.ts:
```diff
@@ -0,0 +1,24 @@
+
+    function create(): ViewBox
+    function add(box: ViewBox, addend: ViewBox): ViewBox
+    function scale(box: ViewBox, scale: number): ViewBox
```
This fails lint in @typescript-eslint/plugin@>=1.4.0, see typescript-eslint/typescript-eslint#316

kasbah commented:
Maybe a future optimization could be to load the SVGs into WebGL (e.g. [with three.js](https://threejs.org/docs/#examples/loaders/SVGLoader))? This could also allow for a 3d view (I noticed I keep wanting to spin the board around in the new tracespace viewer, as I do in the KiCad 3d viewer).

mcous commented:
That's really cool, I haven't kept up with three.js, and I didn't know it could load SVGs now. Definitely down to add a 3D tab
